File: DAQ_Zynq_GUI/SW/Portal/app/adc_pmod_backend.py
# adc_pmod_backend.py
# Only this file touches juliacall for the ADC_PMOD path.
# Julia bootstrap (env vars, jl.include) happens ONCE in app.py, before this
# module is imported — this file assumes juliacall is already usable.

import os
from pathlib import Path
from juliacall import Main as jl

THIS_DIR = os.path.dirname(os.path.abspath(__file__))
jl.include(os.path.join(THIS_DIR, "adc_pmod_backend.jl"))   # renamed from adc_backend_portal.jl

VREF = 3.3
BITS = 12
F_SMPL = 118000          # the ONE real, fixed rate for this hardware — not user-selectable

def capture(ch: int, n: int):
    """Blocking. Must be called on the Qt main thread only."""
    samples = jl.capture(ch, n)
    import numpy as np
    return np.array(samples, dtype=np.uint32)

def is_running(ch: int) -> bool:
    return jl.is_running(ch)

def start(ch: int):
    print(f"Starting ADC_PMOD capture on channel {ch}...")
    jl.start(ch)

def stop(ch: int):
    print(f"Stopping ADC_PMOD capture on channel {ch}...")
    jl.stop(ch)

def t_axis(n):
    import numpy as np
    return np.arange(n) / F_SMPL * 1000.0

File: DAQ_Zynq_GUI/SW/Portal/app/adc_pmod_backend.jl
#!/usr/bin/env julia


const ADC_PMOD_VREF_MV = 3300.0
const ADC_PMOD_BITS = 12
const F_SMPL = 118000

# ── Helpers ───────────────────────────────────────────────────────────────────

adc_to_mv(sample::Integer) = sample * ADC_PMOD_VREF_MV / ((1 << ADC_PMOD_BITS) - 1)
t_axis(n) = collect(0:n-1) ./ F_SMPL .* 1000  # ms

include(joinpath(@__DIR__, "..", "priv", "adc_pmod_ctrl_cfg.jl"))
    
function capture(ch::Int, n::Int)

    if ch == 0
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_0)
    elseif ch == 1
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_1)
    else
        error("Channel must be 0 or 1") 
    end
    
    samples = zeros(UInt32, n)
    try
        cnv_trig(adc, ch, n)
        while true
            sleep(0.1)
            progress = cnv_progress(adc)
            println("Progress: $progress%")
            progress == 100 && break
        end
        read_buf!(adc, ch, samples)
    catch e
        println("Capture error: $e")
        rethrow()          # let caller know it failed
    end
    return adc_to_mv(samples)         # <-- here: returned on success
end

function is_running(ch::Int)::Bool
    
    if ch == 0
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_0)
    elseif ch == 1
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_1)
    else
        error("Channel must be 0 or 1") 
    end
    
    return read_word(adc, adc_pmod_ctrl_cfg.RM_ch_enable) & (1 << ch) != 0
end

function start(ch::Int)
    println("Starting ADC_PMOD capture on channel $ch...")

    if ch == 0
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_0)
    elseif ch == 1
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_1)
    else
        error("Channel must be 0 or 1")
    end

    println("Dummy function, already started in capture()")
end

function stop(ch::Int)
    println("Stopping ADC_PMOD capture on channel $ch...")
    if ch == 0
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_0)
    elseif ch == 1
        adc = ADC_PMOD_CTRL(get_portal(), PG_ADC_PMOD_1)
    else
        error("Channel must be 0 or 1")
    end

    write_word(adc, adc_pmod_ctrl_cfg.RM_ch_enable, 0)
    
end
